### PR TITLE
refactor: audio implementation

### PR DESCRIPTION
**Pull request by Landeplage**
# What?
Refactored the implementation of audio, using Scriptable Objects for all audio events.

# Why?
To prevent issues with git merging while working further on the audio implementation, and for scalability. Also, because scriptable objects are just plain neat.

### CODE REVIEW

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs [hunk 1]
```diff
@@ -62,6 +61,9 @@ public class AvatarEditorWearableFilter
     private AvatarEditorHUDController controller;
     internal readonly Dictionary<string, ItemSelector> selectorsByCategory = new Dictionary<string, ItemSelector>();
 
+    [HideInInspector]
+    public System.Action<AvatarModel> onAvatarAppear;
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs]:
```suggestion
    public event System.Action<AvatarModel> OnAvatarAppear;
```

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs [hunk 2]
```diff
@@ -229,33 +231,8 @@ public void UpdateAvatarPreview(AvatarModel avatarModel)
...
-                        {
-                            if (!eventReaction.source.isPlaying)
-                                eventReaction.PlayScheduled(0.6f);
-                        }
-                    }
-                }
-
-                
+                if (onAvatarAppear != null)
+                    onAvatarAppear.Invoke(avatarModel);
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs]:
```suggestion
                    OnAvatarAppear?.Invoke(avatarModel);
```

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs [hunk 3]
```diff
@@ -133,42 +133,42 @@ public void Populate(Model chatEntryModel)
...
-                        HUDAudioPlayer.i.Play(HUDAudioPlayer.Sound.receiveGlobalChatEntry);
+                        AudioScriptableObjects.chatReceiveGlobal.Play(true);
                         break;
                     default:
                         break;
                 }
             }
 
-            HUDAudioPlayer.i.RefreshChatLastCheckedTimestamp();
+            HUDAudioHandler.i.RefreshChatLastCheckedTimestamp();
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs:171]: I think we can move this `RefreshChatLastCheckedTimestamp` to this class, and remove `HUDAudioHandler` completely. 😄

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs [hunk 4]
```diff
@@ -74,6 +73,9 @@ public bool parcelHighlightEnabled
 
         private bool isInitialized = false;
 
+        [HideInInspector]
+        public System.Action onMovedParcelCursor;
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs]:
```suggestion
        public event System.Action OnMovedParcelCursor;
```

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs [hunk 5]
```diff
@@ -206,9 +208,10 @@ void UpdateParcelHighlight()
             parcelHighlightImage.transform.position = worldCoordsOriginInMap + cursorMapCoords * parcelSizeInMap + new Vector3(parcelSizeInMap, parcelSizeInMap, 0f) / 2;
             highlightedParcelText.text = showCursorCoords ? $"{cursorMapCoords.x}, {cursorMapCoords.y}" : string.Empty;
 
-            if (audioHandler != null && highlightedParcelText.text != previousText && !Input.GetMouseButton(0))
+            if (highlightedParcelText.text != previousText && !Input.GetMouseButton(0))
             {
-                audioHandler.PlayMapParcelHighlight();
+                if (onMovedParcelCursor != null)
+                    onMovedParcelCursor.Invoke();
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs]:
```suggestion
                    OnMovedParcelCursor?.Invoke();
```

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/SliderAudioHandler.cs [hunk 6]
```diff
@@ -4,57 +4,44 @@
...
-    HUDAudioPlayer audioPlayer;
     Slider slider;
-    float valueIncrementTimer = 0f;
+    AudioEvent eventValueChanged;
 
-    void Start()
+    void Awake()
     {
-        audioPlayer = HUDAudioPlayer.i;
+        eventValueChanged = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/SliderValueChange");
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/SliderAudioHandler.cs]: Can't we move this to `AudioScriptableObjects`?

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/ToggleAudioHandler.cs [hunk 7]
```diff
@@ -4,38 +4,44 @@
...
     Toggle toggle;
+    AudioEvent eventEnable, eventDisable;
 
-    void Start()
+    void Awake()
     {
-        audioPlayer = HUDAudioPlayer.i;
+        eventEnable = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/Enable");
+        eventDisable = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/Disable");
+
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/ToggleAudioHandler.cs]: Can't we move those to `AudioScriptableObjects`?

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/Audio/AudioContainer.cs [hunk 8]
```diff
@@ -38,14 +33,12 @@ void Awake()
...
+                Debug.LogWarning("There are no clips in the audio event '" + e.name + "' (" + name + ")");
                 continue;
             }
 
-            e.Initialize();
-
-            // Set up AudioSource component for event
-            e.source = gameObject.AddComponent(typeof(AudioSource)) as AudioSource;
+            // Add AudioSource component for event
+            e.Initialize(gameObject.AddComponent(typeof(AudioSource)) as AudioSource);
```
Comment (BrianAmadori) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/Audio/AudioContainer.cs]: Maybe we can just pass the `AudioContainer` here, and move all the `AddComponent` and `AudioSource` setup inside that method? This way the entire initialization gets encapsulated inside `AudioEvent`.

**Review by BrianAmadori — CHANGES_REQUESTED** [review]
Orgasmic refactor. Left some minor requests to put the cherry on top of this PR. 🍒

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs (hunk 3 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs:171]: Yes! I'm a bit unclear on which of these I should use. I'll experiment and see if I can figure it out.
image

Comment (BrianAmadori) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs:171]: I think you can use the `lastReadWorldChatMessages` one.

**Review by BrianAmadori — APPROVED** [review]
🚀

**Pull request merged**